OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs

```diff
@@ -338,64 +339,87 @@ public bool InterpretCommand(S server, Connection conn, Session.Client client, s
-						var slots = server.LobbyInfo.Slots.Keys.ToArray();
-						var i = 0;
-						foreach (var os in oldSlots)
-						{
-							var c = server.LobbyInfo.ClientInSlot(os);
-							if (c == null)
-								continue;
-
-							c.SpawnPoint = 0;
-							c.Slot = i < slots.Length ? slots[i++] : null;
-							if (c.Slot != null)
+							// Make sure the map hasn't changed in the meantime
+							if (server.LobbyInfo.GlobalSettings.Map != lastMap)
+								return;
+
+							server.LobbyInfo.GlobalSettings.Map = map.Uid;
+
+							var oldSlots = server.LobbyInfo.Slots.Keys.ToArray();
+							server.Map = server.ModData.MapCache[server.LobbyInfo.GlobalSettings.Map];
+
+							server.LobbyInfo.Slots = server.Map.Players.Players
+								.Select(p => MakeSlotFromPlayerReference(p.Value))
+								.Where(ss => ss != null)
+								.ToDictionary(ss => ss.PlayerReference, ss => ss);
+
+							LoadMapSettings(server.LobbyInfo.GlobalSettings, server.Map.Rules);
+
+							// Reset client states
+							foreach (var c in server.LobbyInfo.Clients)
+								c.State = Session.ClientState.Invalid;
+
+							// Reassign players into new slots based on their old slots:
+							//  - Observers remain as observers
+							//  - Players who now lack a slot are made observers
+							//  - Bots who now lack a slot are dropped
+							//  - Bots who are not defined in the map rules are dropped
+							var botNames = server.Map.Rules.Actors["player"].TraitInfos<IBotInfo>().Select(t => t.Name);
+							var slots = server.LobbyInfo.Slots.Keys.ToArray();
+							var i = 0;
+							foreach (var os in oldSlots)
 							{
-								// Remove Bot from slot if slot forbids bots
-								if (c.Bot != null && (!server.Map.Players.Players[c.Slot].AllowBots || !botNames.Contains(c.Bot)))
+								var c = server.LobbyInfo.ClientInSlot(os);
+								if (c == null)
+									continue;
+
+								c.SpawnPoint = 0;
+								c.Slot = i < slots.Length ? slots[i++] : null;
+								if (c.Slot != null)
+								{
+									// Remove Bot from slot if slot forbids bots
+									if (c.Bot != null && (!server.Map.Players.Players[c.Slot].AllowBots || !botNames.Contains(c.Bot)))
+										server.LobbyInfo.Clients.Remove(c);
+									S.SyncClientToPlayerReference(c, server.Map.Players.Players[c.Slot]);
+								}
+								else if (c.Bot != null)
 									server.LobbyInfo.Clients.Remove(c);
-								S.SyncClientToPlayerReference(c, server.Map.Players.Players[c.Slot]);
 							}
-							else if (c.Bot != null)
-								server.LobbyInfo.Clients.Remove(c);
-						}
 
-						// Validate if color is allowed and get an alternative it isn't
-						foreach (var c in server.LobbyInfo.Clients)
-							if (c.Slot == null || (c.Slot != null && !server.LobbyInfo.Slots[c.Slot].LockColor))
-								c.Color = c.PreferredColor = SanitizePlayerColor(server, c.Color, c.Index, conn);
+							// Validate if color is allowed and get an alternative it isn't
```

Comment:
I don't understand your comment here...?  In any case, this is just moving existing code.